pkg/progress/progress.go

```diff
@@ -62,52 +83,89 @@ func (m *Manager) Reset() {
+type Option func(*progressIndicator)
+
+// WindowDurationOption changes the time window size.
+func WindowDurationOption(dur time.Duration) func(*progressIndicator) {
+	return func(pi *progressIndicator) {
+		if dur < minSpeedCalculationWindow {
+			dur = minSpeedCalculationWindow
+		} else if dur > maxSpeedCalculationWindow {
+			dur = maxSpeedCalculationWindow
+		}
+		pi.windowLength = int(dur/pi.updateInterval) + 1
+	}
+}
+
 // AddProgress adds a progress into manager if it doesn't exist.
-func (m *Manager) AddProgress(progress string, current, total float64, updateInterval time.Duration) (exist bool) {
+func (m *Manager) AddProgress(progress string, current, total float64, updateInterval time.Duration, opts ...Option) (exist bool) {
 	m.Lock()
 	defer m.Unlock()
 
 	history := list.New()
 	history.PushBack(current)
 	if _, exist = m.progresses[progress]; !exist {
-		m.progresses[progress] = &progressIndicator{
-			total:             total,
-			remaining:         total,
-			history:           history,
-			windowLengthLimit: int(speedStatisticalWindow / updateInterval),
-			updateInterval:    updateInterval,
+		pi := &progressIndicator{
+			total:          total,
+			remaining:      total,
+			history:        history,
+			windowCapacity: int(maxSpeedCalculationWindow/updateInterval) + 1,
+			windowLength:   int(minSpeedCalculationWindow / updateInterval),
+			updateInterval: updateInterval,
+		}
+		for _, op := range opts {
+			op(pi)
 		}
+		m.progresses[progress] = pi
+		pi.front = history.Front()
+		pi.currentWindowLength = 1
 	}
 	return
 }
 
 // UpdateProgress updates the progress if it exists.
-func (m *Manager) UpdateProgress(progress string, current, remaining float64, isInc bool) {
+func (m *Manager) UpdateProgress(progress string, current, remaining float64, isInc bool, opts ...Option) {
 	m.Lock()
 	defer m.Unlock()
 
 	if p, exist := m.progresses[progress]; exist {
+		for _, op := range opts {
+			op(p)
+		}
 		p.remaining = remaining
 		if p.total < remaining {
 			p.total = remaining
 		}
 
-		if p.history.Len() > p.windowLengthLimit {
+		p.history.PushBack(current)
+		p.currentWindowLength++
+
+		// try to move `front` into correct place.
+		for p.currentWindowLength > p.windowLength {
+			p.front = p.front.Next()
+			p.currentWindowLength--
+		}
+		for p.currentWindowLength < p.windowLength && p.front.Prev() != nil {
+			p.front = p.front.Prev()
+			p.currentWindowLength++
+		}
+
+		for p.history.Len() > p.windowCapacity {
 			p.history.Remove(p.history.Front())
 		}
-		p.history.PushBack(current)
 
 		// It means it just init and we haven't update the progress
 		if p.history.Len() <= 1 {
 			p.lastSpeed = 0
 		} else if isInc {
 			// the value increases, e.g., [1, 2, 3]
-			p.lastSpeed = (p.history.Back().Value.(float64) - p.history.Front().Value.(float64)) /
-				(float64(p.history.Len()-1) * p.updateInterval.Seconds())
+			p.lastSpeed = (p.history.Back().Value.(float64) - p.front.Value.(float64)) /
```

Comment:
can we use `current` to replace `p.history.Back()`?

Reply:
Yes